### ISSUE TITLE
iOS initial map update patch

### ISSUE THREAD

**Issue by cheolgi** [tangrams/tangram-es#964]
To whom it may concern,

I am Cheolgi Kim, who started to make an app using tangarm-es, recently.

I found that the map does not appear initially in IOS before I touch the screen to activate a new rendering. The reason was because setupGL() calls loadSceneAsync(), which pends some remained work for the UI task into jobQueue. In OSX, render() method, which processes pending work in jobQueue, is called eagerly in the main loop. However, iOS version does not call render() lazily (maybe because of battery life).

I made a patch to process jobQueue eagearly in the background GUI task ASAP using dispatch_async() function. This patch let the iOS app render the map when it starts while using loadSceneAsync().

I hope this contribute a little to the project.

==
```
diff --git a/core/src/platform.h b/core/src/platform.h
index 7fc6497..95b50aa 100644
--- a/core/src/platform.h
+++ b/core/src/platform.h
@@ -127,3 +127,11 @@ do { logMsg("TANGRAM %s:%d: " fmt "\n", __FILENAME__, __LINE__, ## __VA_ARGS__);
 
 #define LOGS(fmt, ...) \
 do { TextDisplay::Instance().log(fmt, ## __VA_ARGS__); } while(0)
+
+/* ADDED BY CHEOLGI */
+#ifdef PLATFORM_IOS
+#include "util/jobQueue.h"
+
+void asyncRunQueue(Tangram::JobQueue* queue);
+#endif
+/* END: ADDED BY CHEOLGI */
diff --git a/core/src/util/jobQueue.cpp b/core/src/util/jobQueue.cpp
index d69b437..8bb2be2 100644
--- a/core/src/util/jobQueue.cpp
+++ b/core/src/util/jobQueue.cpp
@@ -1,4 +1,9 @@
 #include "jobQueue.h"
+/* ADDED BY CHEOLGI */
+#ifdef PLATFORM_IOS
+#include "platform.h"
+#endif
+/* END OF ADDED BY CHEOLGI */
 
 namespace Tangram {
 
@@ -15,6 +20,11 @@ void JobQueue::add(Job job) {
     std::lock_guard<std::mutex> lock(m_mutex);
     m_jobs.push_back(job);
 
+    /* ADDED BY CHEOLGI */
+#ifdef PLATFORM_IOS
+    asyncRunQueue(this);
+#endif
+    /* END: ADDED BY CHEOLGI */
 }
 
 void JobQueue::runJobs() {
diff --git a/ios/src/TGMapViewController.mm b/ios/src/TGMapViewController.mm
index 97c3e42..b1e0184 100644
--- a/ios/src/TGMapViewController.mm
+++ b/ios/src/TGMapViewController.mm
@@ -342,13 +342,21 @@ - (void)didReceiveMemoryWarning
     // Dispose of any resources that can be recreated.
 }
 
+static void requestRenderAsync()
+{
+    dispatch_async(dispatch_get_main_queue(), ^{  
+            LOG("READY");
+            requestRender();
+        });  
+}
+
 - (void)setupGL
 {
     [EAGLContext setCurrentContext:self.context];
 
     if (!self.map) {
         self.map = new Tangram::Map();
-        self.map->loadSceneAsync("scene.yaml");
+        self.map->loadSceneAsync("scene.yaml", false,requestRenderAsync);
     }
     self.map->setupGL();
 
diff --git a/ios/src/platform_ios.mm b/ios/src/platform_ios.mm
index b467055..b0e80f1 100644
--- a/ios/src/platform_ios.mm
+++ b/ios/src/platform_ios.mm
@@ -10,6 +10,8 @@
 #include "platform_ios.h"
 #include "TGMapViewController.h"
 
+#include "util/jobQueue.h"
+
 static TGMapViewController* viewController;
 NSURLSession* defaultSession;
 
@@ -31,6 +33,14 @@ void init(TGMapViewController* _controller) {
 
 }
 
+/* CHEOLGI ADDED */
+void asyncRunQueue(Tangram::JobQueue* queue) {
+  dispatch_async(dispatch_get_main_queue(), ^{
+    queue->runJobs();
+  });
+}
+/* END OF CHEOLGI ADDED */
+
 void logMsg(const char* fmt, ...) {
 
     va_list args;
```

**Comment by blair1618** [comment]
Hi Cheolgi, thanks for the information and the patch! I've noticed a similar issue in the OS X and Android demo applications (https://github.com/tangrams/tangram-es/issues/963) and I'll consider this when debugging those as well.

**Comment by cheolgi** [comment]
I think this issue is implicitly merged to [#963](https://github.com/tangrams/tangram-es/issues/963). I think it can be closed.
